File: scripts/export.lua
-- only supports Catppuccin flavors

local color_names = {
    "rosewater", "flamingo", "pink", "mauve", "red", "maroon", "peach",
    "yellow", "green", "teal", "sky", "sapphire", "blue", "lavender",
    "text", "subtext1", "subtext0", "overlay2", "overlay1", "overlay0",
    "surface2", "surface1", "surface0", "base", "mantle", "crust"
}

local function read_file(path)
    local file = io.open(path, "r")
    if not file then
        return nil
    end

    local content = file:read("*a")
    file:close()
    return content
end

local function write_file(path, content)
    local file = assert(io.open(path, "w"))
    file:write(content)
    file:close()
end

local function valid_name(name)
    return name and name:match("^[%w_-]+$") ~= nil
end

local function read_path(prompt)
    io.write(prompt)
    io.flush()
    return io.read("*l")
end

local function extract_colors(css)
    local colors = {}

    for name, value in css:gmatch("%-%-([%w_-]+)%s*:%s*(#[%da-fA-F]+)%s*;") do
        colors[name] = value
    end

    return colors
end

local function json_escape(value)
    return value:gsub('\\', '\\\\'):gsub('"', '\\"')
end

local function export_theme()
    print("colorscheme to export: cfrappe, clatte, cmocha, dracula or nord")
    local colorscheme = io.read()

    if colorscheme == "nord" or colorscheme == "dracula" then
        print("not supported yet")
        return
    end

    if not valid_name(colorscheme) then
        print("write cfrappe, clatte, cmocha, dracula or nord")
        return
    end

    local destination = read_path("where should exported.jsonc be saved? ")
    if not destination or destination == "" then
        print("destination is required")
        return
    end

    local css = read_file("src/" .. colorscheme .. "/trans.css")
    if not css then
        print("colorscheme does not exist")
        return
    end

    local colors = extract_colors(css)
    local output = '{\n  "name": "' .. json_escape(colorscheme) .. '",\n  "colors": {\n'

    for index, name in ipairs(color_names) do
        local value = colors[name]
        if not value then
            print("colorscheme is missing --" .. name)
            return
        end

        output = output .. '    "' .. name .. '": "' .. value .. '"'
        if index < #color_names then
            output = output .. ","
        end
        output = output .. "\n"
    end

    output = output .. "  }\n}\n"
    local output_path = destination .. "/exported.jsonc"
    write_file(output_path, output)
    print("exported to " .. output_path)
end

export_theme()

File: scripts/export.py
#!/usr/bin/env python3

import json
import re
from pathlib import Path


COLOR_NAMES = [
    "rosewater", "flamingo", "pink", "mauve", "red", "maroon", "peach",
    "yellow", "green", "teal", "sky", "sapphire", "blue", "lavender",
    "text", "subtext1", "subtext0", "overlay2", "overlay1", "overlay0",
    "surface2", "surface1", "surface0", "base", "mantle", "crust",
]


def read_file(path: Path) -> str | None:
    try:
        return path.read_text()
    except FileNotFoundError:
        return None


def valid_name(name: str | None) -> bool:
    return bool(name and re.fullmatch(r"[\w_-]+", name))


def extract_colors(css: str) -> dict[str, str]:
    colors: dict[str, str] = {}
    for name, value in re.findall(r"--([\w_-]+)\s*:\s*(#[\da-fA-F]+)\s*;", css):
        colors[name] = value
    return colors


def export_theme() -> None:
    print("colorscheme to export: cfrappe, clatte, cmocha, dracula or nord")
    colorscheme = input().strip()

    if colorscheme in {"nord", "dracula"}:
        print("not supported yet")
        return

    if not valid_name(colorscheme):
        print("write cfrappe, clatte, cmocha, dracula or nord")
        return

    destination = input("where should exported.jsonc be saved? ").strip()
    if not destination:
        print("destination is required")
        return

    css = read_file(Path("src") / colorscheme / "trans.css")
    if not css:
        print("colorscheme does not exist")
        return

    colors = extract_colors(css)
    output = {"name": colorscheme, "colors": {}}

    for name in COLOR_NAMES:
        value = colors.get(name)
        if not value:
            print(f"colorscheme is missing --{name}")
            return
        output["colors"][name] = value

    output_path = Path(destination) / "exported.jsonc"
    output_path.write_text(json.dumps(output, indent=2) + "\n")
    print(f"exported to {output_path}")


if __name__ == "__main__":
    export_theme()
